packages/core/src/test/codewhispererChat/editor/codelens.test.ts

```diff
@@ -0,0 +1,72 @@
+/*!
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import assert from 'assert'
+import vscode from 'vscode'
+import {
+    TryChatCodeLensProvider,
+    resolveModifierKey,
+    tryChatCodeLensCommand,
+} from '../../../codewhispererChat/editor/codelens'
+import { installFakeClock } from '../../testUtil'
+import { InstalledClock } from '@sinonjs/fake-timers'
+import globals from '../../../shared/extensionGlobals'
+
+describe('TryChatCodeLensProvider', () => {
+    let instance: TryChatCodeLensProvider = new TryChatCodeLensProvider()
+    let cancellationTokenSource: vscode.CancellationTokenSource
+    let clock: InstalledClock
+    const codeLensPosition = new vscode.Position(1, 2)
+
+    beforeEach(function () {
+        instance = new TryChatCodeLensProvider(() => codeLensPosition)
+        clock = installFakeClock()
+    })
+
+    afterEach(function () {
+        instance.dispose()
+        cancellationTokenSource?.dispose()
+        clock.uninstall()
+    })
+
+    it('keeps returning a code lense until it hits the max times it should show', async function () {
+        let codeLensCount = 0
+        const modifierKey = resolveModifierKey()
+        while (codeLensCount < TryChatCodeLensProvider.maxCount) {
+            cancellationTokenSource = new vscode.CancellationTokenSource()
+            const resultPromise = instance.provideCodeLenses({} as any, cancellationTokenSource.token)
+            clock.tick(TryChatCodeLensProvider.debounceMillis) // skip debounce
+
+            assert.deepStrictEqual(await resultPromise, [
+                {
+                    range: new vscode.Range(codeLensPosition, codeLensPosition),
+                    command: {
+                        title: `Amazon Q: explain in chat (${modifierKey} + i) - showing ${
+                            TryChatCodeLensProvider.maxCount - codeLensCount
+                        } more times`,
+                        command: tryChatCodeLensCommand.id,
+                        arguments: [instance],
+                    },
+                    isResolved: true,
+                },
+            ])
+
+            codeLensCount++
+        }
+        const emptyResult = await instance.provideCodeLenses({} as any, new vscode.CancellationTokenSource().token)
+        assert.deepStrictEqual(emptyResult, [])
+    })
+
+    it('does not register the provider if we do not want to show the code lens', async function () {
+        await globals.context.globalState.update(TryChatCodeLensProvider.showCodeLensId, true)
```

Comment:
Should globals be mocked so that it doesn't effect the actual runtime global state? (or does the test harness already do this?)